src/dask_awkward/lib/io/io.py

```diff
@@ -565,21 +582,20 @@ def from_map(
     if io_func_implements_mocking(func):
         io_func = func
         array_meta = cast(ImplementsMocking, func).mock()
-    # If we know the meta, we can spoof mocking
+    # # If we know the meta, we can spoof mocking
     elif meta is not None:
-        io_func = IOFunctionWithMocking(meta, func)
+        # io_func = IOFunctionWithMocking(meta, func)
+        io_func = func
```

Comment:
@agoose77 - what was the need for this part of the `from_map` code (wrapping `io_func` in `IOFunctionWithMocking`). I ran into a problem here when writing a test for the new feature of this PR. If `mock` wasn't defined but `meta` was, If I tried to use `self.form` in the `io_func` that I defined, I lost the ability to get the `form` attr from my `io_func` because it was now wrapped by `IOFunctionWithMocking` which only has two attrs: `_io_func` and and `_meta`. I removed this problem by utilizing `mock` (as you suggested!) as part of this PR, but it was still a confusing thing to run into to see the `io_func` that I wrote get covered up by something else.

I'm also thinking it's unnecessary to bolt on an available `mock` method if it's not already there. If consuming code wants to support necessary-data-projection that consuming code will definitely need an `io_func` with a `mock` method

Reply:
This was written from first-principles (and to support the existing API), rather than as a use-case.

I think it addresses the case that there are _two_ input layers; one that supports column optimisation, and the other which does not. By allowing `meta` to be used during `mock`, we can avoid computing the non-optimising input during optimisation of the graph. 

This might be useful for users who cannot selectively read but do know the type of their data up-front.

As such, I think it's an important feature

Reply:
Thanks for the explanation! That does make sense w.r.t. multiple (mixed source) input layers. It would be quite strict to require all io funcs to be function classes (for example `read_text` just uses a standalone function and derives a meta, so it would use this feature if combined with another optimizable input layer). Seems like things should stay as-is